### DISCUSSION TITLE
Learning an expression with four terms

### DISCUSSION THREAD

**Discussion by strifinopoulos** [MilesCranmer/PySR#240] [Q&A]
I am trying to learn an equation of the form: 

a*(x+y)+b*(x+y)^(2/3)-c*x*(x-1)/(x+y)^(1/3)-d*(y-x)^2/(x+y)

where a,b,c,d are constants

Things I have tried: 
-Change variables to x, z=x+y, w=x-y and train on them
-define: ()^(2/3), ()^(1/3) and square as new unary operators (and put constraints on their complexities <3)
-increase d the maxdepth=maxsize=30
-increase the population_size
-increase the complexity of constants
-randomize the tree more frequently
-let the code run until loss<10

I am running out of ideas and the loss is still > 300 (the program halts since there is no improvement)
I would be grateful for any ideas.

**Comment by MilesCranmer** [comment]
What's the range of x and y?

Regardless, this looks like a pretty difficult expression to find. But it might be possible.

**Reply by strifinopoulos** [reply]
x and y are positive integer numbers.  Is there any way to specify that there are four terms? that might be a good start.

**Reply by MilesCranmer** [reply]
One thing you could do is play around with the `nested_constraints` and `constraints` so that only `+` and `-` can be nested repeatedly. That will essentially force it to have a sum of simple terms.

**Reply by MilesCranmer** [reply]
The other option is to use this PR: https://github.com/MilesCranmer/SymbolicRegression.jl/pull/143 which is not yet merged, and heed the comment [here](https://github.com/MilesCranmer/SymbolicRegression.jl/pull/143#issuecomment-1279205103). 

> One easy way to get multiple expressions out is to enforce that the root node is a binary operator (or else, an infinite loss is returned), and is ignored. Then, `tree.l` is ϕ and `tree.r` is ψ, and you can do your evaluation on each, using separate calls to `eval_tree_array`. Sometimes it might delete the right child, but that's okay since those branches of the evolutionary tree would just die out, given their large loss.

You would basically have `tree.l.l` be term 1, `tree.l.r` be term 2, `tree.r.l` be term 3, and `tree.r.r` be term 4. You would to double-check that the tree is composed of two binary operators at the top levels though, before using `tree.l.l`.

However, you'll have to use the Julia backend to configure this, rather than PySR. Once I get around to merging it I will think about how to make it configurable in PySR though.

**Reply by MilesCranmer** [reply]
Also see https://astroautomata.com/PySR/backend/ for customizing the backend for use in PySR.
